docs/current/guides/757394-use-services.md

```diff
@@ -0,0 +1,347 @@
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+Here's what happens on the last line:
+
+1. The client requests the `ping` container's stdout, which requires the container to run.
+1. Dagger sees that the `ping` container has a service binding, `redisSrv`.
+1. Dagger starts the `redisSrv` container, which recurses into this same process.
+1. Dagger waits for health checks to pass against `redisSrv`.
+1. Dagger runs the `ping` container with the `redis-srv` alias magically added to `/etc/hosts`.
+
+:::note
+Dagger cancels each service run after a 10 second grace period to avoid frequent restarts.
+:::
+
+It's worth noting that services are just containers, and all containers in Dagger have run-exactly-once semantics. Concurrent runs of the same container synchronize and attach to the same run with progress/logs multiplexed to each caller. Canceling a run only interrupts the container process when all runs are canceled, so new clients can come and go throughout a service run.
+
+Run-exactly-once semantics are very convenient. You don't have to come up with names and maintain instances of services; they're content-addressed, so you just use them by value. You also don't have to manage the state of the service; you can just trust that it will be running when needed and stopped when not.
+
+:::tip
+If you need multiple instances of a service, just attach something unique to each one, such as an instance ID.
+:::
+
+Let's put all this together in a full client-server example of running commands against a Redis service:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="d4SIgAW2Feo" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+Note that this example relies on the 10-second grace period, which you should try to avoid. It would be better to chain both commands together, which ensures that the service stays running for both:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="avO1QVAIwBZ" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+## Persisting service state
+
+Another way to avoid relying on the grace period is to use a cache volume to persist a service's data, as in the following example:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="23lXKbJiCz0" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+Note that this example uses Redis's `SAVE` command to ensure data is synced. By default, Redis flushes data to disk periodically.
+
+## Example: Provisioning and testing a Redis service
+
+The following composite example uses Dagger to provision a Redis server with data persisted to a named cache and runs a user-provided command against it.
```

Comment:
@vito could you add some context ie. what is this example intended to demonstrate?

Reply:
TBH I threw this in at the end but was thinking about removing it. I think it would be better to have a more realistic example, like running integration tests with access to a database. I know people do this a lot but don't have an example within reach at the moment.

Reply:
@vito so should we leave this in as is or remove it?

Reply:
Deleted